File: level06/Ressources/keygen.py
#!/usr/bin/env python
# Level06 Serial Number Keygen

def calculate_serial(login):
    """Calculate the correct serial number for a given login."""
    
    # Remove newline if present
    login = login.strip()
    
    # Check constraints
    if len(login) <= 5:
        print("Error: Login must be more than 5 characters")
        return None
    
    for c in login:
        if ord(c) <= 31:
            print("Error: All characters must be printable (ASCII > 31)")
            return None
    
    # Calculate serial using the same algorithm as auth()
    v4 = (ord(login[3]) ^ 0x1337) + 6221293
    
    for i in range(len(login)):
        v4 += (v4 ^ ord(login[i])) % 0x539
    
    return v4

if __name__ == "__main__":
    import sys
    
    if len(sys.argv) > 1:
        login = sys.argv[1]
    else:
        login = raw_input("Enter login: ")
    
    serial = calculate_serial(login)
    if serial is not None:
        print("Serial: %d" % serial)

File: level06/Ressources/keygen.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

unsigned int calculate_serial(const char *login) {
    size_t len = strlen(login);

    if (len < 4) {
        fprintf(stderr, "Login must be at least 4 characters\n");
        exit(1);
    }

    unsigned int v4 = ((unsigned int)login[3] ^ 0x1337) + 6221293;

    for (size_t i = 0; i < len; i++) {
        v4 += (v4 ^ (unsigned int)login[i]) % 0x539;
    }

    return v4;
}

int main() {
    char login[256];

    printf("Enter login: ");
    fgets(login, sizeof(login), stdin);

    login[strcspn(login, "\n")] = '\0';

    unsigned int serial = calculate_serial(login);
    printf("Serial: %u\n", serial);

    return 0;
}


